src/contrasts.jl

```diff
@@ -466,49 +474,79 @@ expressed as a hypothesis matrix, but it is not obvious just from looking at the
 contrasts matrix.  For this reason `HypothesisCoding` is preferred for
 specifying custom contrast coding schemes over `ContrastsCoding`.
 
+Optional arguments `levels` and `labels` give the ordering of mapping between
+levels and columns of the hypothesis matrix, and the labels for the generated
+contrasts, respetively.
 """
-mutable struct HypothesisCoding <: AbstractContrasts
-    hypotheses::Matrix
-    contrasts::Matrix
-    base::Nothing
-    levels::Union{Vector,Nothing}
-
-    function HypothesisCoding(hypotheses, base, levels)
+mutable struct HypothesisCoding{T<:AbstractMatrix, S<:AbstractMatrix} <: AbstractContrasts
+    hypotheses::T
+    contrasts::S
+    levels::Union{AbstractVector,Nothing}
+    labels::AbstractVector
+
+    function HypothesisCoding(hypotheses::T, levels, labels) where {T}
+        labels == nothing &&
+            throw(ArgumentError("must specify contrast labels with " *
+                                "HypothesisCoding(; labels=) or " *
+                                "HypothesisCoding(Dict(label1=>hyp1, label2=>hyp2, ...))"))
         contrasts = pinv(hypotheses)
+        S = typeof(contrasts)
         check_contrasts_size(contrasts, levels)
-        new(hypotheses, contrasts, base, levels)
+        new{T,S}(hypotheses, contrasts, levels, labels)
     end
 end
 
-HypothesisCoding(mat::Matrix; levels=nothing) =
-    HypothesisCoding(mat, nothing, levels)
+HypothesisCoding(mat::AbstractMatrix; levels=nothing, labels=nothing) =
+    HypothesisCoding(mat, levels, labels)
+
+"""
+    HypothesisCoding(hypotheses::Dict[; labels=, levels=])
+
+Specify hypotheses as `label=>hypothesis_vector` pairs.  If labels are specified
+via keyword argument, the hypothesis vectors will be concatenated in that order.
+"""
+function HypothesisCoding(
+    hypotheses::Dict{<:Any,<:AbstractVector};
+    labels=collect(keys(hypotheses)),
+    levels=nothing
+)
+    !isempty(symdiff(keys(hypotheses), labels)) &&
+        throw(ArgumentError("labels mismatch between hypotheses " *
+                            "($(collect(keys(hypotheses)))) " *
+                            "and labels kw arg ($labels)"))
+    
+    mat = reduce(hcat, hypotheses[label] for label in labels)'
```

Comment:
`reduce(hcat, ...)` is only optimized for vector inputs. So you need to choose between allocating a copy of the input, and using the slow `reduce` which makes N copies.
```suggestion
    mat = reduce(hcat, collect(hypotheses[label] for label in labels))'
```